### PR TITLE
fix: route __call__ through nn.Module to preserve hooks

### LINKED ISSUE #135: [Issue]: __call__ override across losses/models, *args/**kwargs tightening
### Type of Issue

Bug Report

### Description

BaseLoss.__call__, BaseModel.__call__, BaseContrastiveDivergence.__call__ all return self.forward(...) directly. This bypasses nn.Module._call_impl

### Steps to Reproduce / Context

1.
2.
3.

### Minimal Code Example (if applicable)

```python

```

### Environment

_No response_

### PR DESCRIPTION
**Pull request by Giggitycountless**
## Summary

Both `BaseLoss.__call__` and `BaseScoreMatching.__call__` were calling `self.forward()` directly, bypassing `nn.Module._call_impl`. This meant PyTorch hooks (forward hooks, backward hooks, etc.) were silently skipped whenever a loss module was invoked.

## Changes

Changed both methods to call `super().__call__()` which goes through the standard `nn.Module.__call__` → `_call_impl` → `forward()` path, preserving all hook machinery while keeping the existing dtype/device conversion logic.

**Files modified:** `torchebm/core/base_loss.py`
- `BaseLoss.__call__` (line 78): `self.forward(...)` → `super().__call__(...)`
- `BaseScoreMatching.__call__` (line 506): `self.forward(...)` → `super().__call__(...)`

## Why this matters

`nn.Module._call_impl` is responsible for:
- Running registered forward pre-hooks and forward hooks
- Setting up `torch.autograd.grad_mode` correctly
- Calling `_slow_forward` for tracing compatibility
- Error handling for missing `forward()` implementations

By bypassing it, any user code that registers hooks on loss modules (e.g., for logging, gradient monitoring, debugging) would silently have those hooks ignored.

## How to test

```python
import torch
from torchebm.core import HarmonicModel
from torchebm.losses import ContrastiveDivergence
from torchebm.samplers import LangevinDynamics

model = HarmonicModel(k=1.0)
sampler = LangevinDynamics(model, step_size=0.01)
loss_fn = ContrastiveDivergence(model=model, sampler=sampler, k_steps=5)

# Verify hooks fire
hook_called = []
loss_fn.register_forward_hook(lambda m, i, o: hook_called.append(True))
x = torch.randn(16, 2)
loss, neg = loss_fn(x)
assert len(hook_called) == 1, f"Expected hook to fire once, got {len(hook_called)}"
```

All existing tests pass with this change.

Fixes #135

### CODE REVIEW

**Comment by soran-ghaderi** [comment]
Thanks for catching this @Giggitycountless. It's clean and passes the tests. it was invisible until someone hits it under DDP or a profiler. 

will merge it as is. just quick heads up on `BaseModel.__call__` as well. it has the same override pattern. I'll open a small follow-up to extend the same pattern there.

once hooks are routed correctly via super(), the .to()-coercing override layer is largely redundant since subclass forward methods already do conditional device coercion, so the follow-up will probably just drop the overrides entirely. I'll co-author you on it.

Thanks again!

**Review by soran-ghaderi — APPROVED** [review]
it's clean and passes the tests.

thank you!
